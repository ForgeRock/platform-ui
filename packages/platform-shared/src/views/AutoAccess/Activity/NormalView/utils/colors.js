/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

export const colorsHex = ['#2ed47a', '#ffb946', '#109cf1', '#f7685b', '#61e2e2'];
export const colorsReadable = ['green', 'yellow', 'blue', 'red', 'cyan'];
