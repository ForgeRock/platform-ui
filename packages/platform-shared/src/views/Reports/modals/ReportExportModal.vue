<!-- Copyright (c) 2023-2025 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <BModal
    data-testid="export-modal"
    id="export-modal"
    ok-only
    ok-variant="outline-primary"
    no-close-on-backdrop
    no-close-on-esc
    size="lg"
    title-class="h5"
    title-tag="h2"
    :ok-title="$t('common.close')"
    :static="isTesting"
    :title="heading"
    @show="emit('show')">
    <template #modal-title>
      <small class="d-block modal-heading">
        {{ heading }}
      </small>
      <div>
        {{ dayjs(data.date).format('MM/DD/YYYY h:mm A') }}
      </div>
    </template>
    <BContainer
      fluid>
      <BRow>
        <BCol
          md="10"
          offset-md="1"
          class="text-center">
          <div class="d-flex justify-content-center align-items-center flex-column fr-report-export-modal-content">
            <FrSpinner
              v-if="props.status === 'exporting'"
              class="m-4"
              testid="spinner-large" />
            <BImg
              v-else-if="props.status === 'error'"
              class="m-4"
              data-testid="fr-export-error-image"
              height="160"
              width="160"
              :src="require('@forgerock/platform-shared/src/assets/images/alert-circle-outline.svg')"
              :alt="$t('common.alertIconAltText')" />
            <BImg
              v-else
              class="m-4"
              height="160"
              width="160"
              :src="require('@forgerock/platform-shared/src/assets/images/check.svg')"
              :alt="$t('common.checkIconAltText')" />
            <h2 class="h4 mb-4">
              {{ modalTitle }}
            </h2>
            <p class="mb-4 text-break">
              {{ modalMessage }}
            </p>
            <FrButtonWithSpinner
              v-if="props.status === 'download' || props.status === 'downloading'"
              class="mb-4"
              data-testid="fr-history-export-download-button"
              icon="file_download"
              variant="primary"
              :button-text="downloadButtonText"
              :disabled="props.status === 'downloading'"
              :show-spinner="props.status === 'downloading'"
              :spinner-text="downloadButtonText"
              @click="emit('download-report', {
                fileType: props.fileType,
                item: props.data,
                exportStatus: props.status
              });" />
          </div>
        </BCol>
      </BRow>
    </BContainer>
  </BModal>
</template>

<script setup>
import { computed } from 'vue';
import dayjs from 'dayjs';
import {
  BCol,
  BContainer,
  BImg,
  BModal,
  BRow,
} from 'bootstrap-vue';
import FrButtonWithSpinner from '@forgerock/platform-shared/src/components/ButtonWithSpinner/';
import FrSpinner from '@forgerock/platform-shared/src/components/Spinner/';
import i18n from '@/i18n';

const props = defineProps({
  data: {
    type: Object,
    default: () => ({}),
  },
  fileType: {
    type: String,
    default: '',
  },
  isTesting: {
    type: Boolean,
    default: false,
  },
  status: {
    type: String,
    default: 'exporting',
    validator(value) {
      return ['exporting', 'download', 'downloading', 'error'].includes(value);
    },
  },
});

const emit = defineEmits(['download-report', 'show']);
const heading = computed(() => i18n.global.t('reports.modal.heading', { fileType: props.fileType }));
const modalTitle = computed(() => {
  switch (props.status) {
    case 'error':
      return i18n.global.t('reports.modal.errorTitle', { fileType: props.fileType });
    case 'exporting':
      return i18n.global.t('reports.modal.title', { fileType: props.fileType });
    default:
      return i18n.global.t('reports.modal.success.title');
  }
});

const modalMessage = computed(() => {
  switch (props.status) {
    case 'error':
      return i18n.global.t('reports.modal.errorMessage');
    case 'exporting':
      return i18n.global.t('reports.modal.exportReportMessage');
    case 'download':
    case 'downloading':
      return i18n.global.t('reports.modal.success.subtitle');
    default:
      return i18n.global.t('reports.modal.errorMessage');
  }
});

const downloadButtonText = computed(() => {
  switch (props.status) {
    case 'downloading':
      return i18n.global.t(i18n.global.t('reports.downloading'));
    default:
      return i18n.global.t('reports.modal.downloadReport', { fileType: props.fileType });
  }
});
</script>

<style scoped>
  .fr-report-export-modal-content {
    height: 400px;
  }

  .modal-heading {
    height: 30px;
  }
</style>
