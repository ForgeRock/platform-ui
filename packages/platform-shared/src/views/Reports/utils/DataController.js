/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import {
  getAMTreesConfig,
  getCertificationsForReports,
  getGovernanceResource,
  getGovPolicyList,
  getGovPolicyRuleList,
  getOauth2Clients,
  getRequestTypesForReports,
  getTeamMembers,
  managedResourcePropertyRequest,
  requestTrees,
} from './ReportsApiHelper';

/**
 * @description Data controller object that maps the data source to the API
 * request. The deeper the level, the more priority it has over the parent
 * level. For example, if a parent level has a fetch function, but the child
 * level also has a fetch function, the child level fetch function will be used.
 *
 * { SERVICE } First Level is the datasource service name (AM, IDM, IGA).
 * { entity } Second Level is the data source entity name.
 * { attribute } Third Level is the data source entity property or attribute name.
 *
 * AVAILABLE PROPERTIES:
 * - fetch: The API request function to fetch the data.
 * - mutation: The function to transform the data.
 * - canQuery: Boolean to determine if the data can be queried.
 * - endUserCanFetch: Boolean to determine if the end user can fetch the data.
 * - taggable: Boolean to determine if the parameter field can be tagged.
 * - {*}: Additional properties can be added and will be injected into
 *   the parameter request object, which in-turn gets injected into the
 *   fetch function as the first argument.
 */

const DATA_CONTROLLER = Object.freeze({
  AM: {
    amconfig: {
      fetch: getAMTreesConfig,
      mutation: ({ suspendedAuthenticationTimeout: first, authenticationSessionsMaxDuration: second }) => Math.min(first, second),
    },
    trees: {
      fetch: requestTrees,
      canQuery: false,
    },
  },
  IDM: {
    fetch: managedResourcePropertyRequest,
    application: {},
    group: {},
    oauth2application: { fetch: getOauth2Clients },
    organization: {},
    role: {},
    teammember: { fetch: getTeamMembers },
    user: {},
  },
  IGA: {
    certification: { fetch: getCertificationsForReports },
    // Currently, IGA does not have the capability to query
    // the entitlements object directly, so the only way to
    // query entitlements is by "displayName" under the
    // "descriptor" object, and this is why "targetName"
    // and "sortBy" are hard-coded to point to "displayName".
    entitlement: {
      fetch: getGovernanceResource,
      mutation: (data, attribute) => data.map((obj) => obj.entitlement[attribute] || obj),
      body: (term = '') => ({
        targetFilter: {
          operator: 'CONTAINS',
          operand: {
            targetName: 'descriptor.idx./entitlement.displayName',
            targetValue: term,
          },
        },
      }),
      queryParameters: {
        pageNumber: 0,
        pageSize: 10,
        sortBy: 'descriptor.idx./entitlement.displayName',
        sortDir: 'asc',
      },
    },
    policy: { fetch: getGovPolicyList },
    request: { fetch: getRequestTypesForReports },
    rule: { fetch: getGovPolicyRuleList },
    violation: {},
  },
});

export default DATA_CONTROLLER;
