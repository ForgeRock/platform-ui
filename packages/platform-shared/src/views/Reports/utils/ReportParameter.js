/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

/**
 * @description Class that generates an analytics report parameter object.
 */

export default class ReportParameter {
  /**
   * @description Constructor for the ReportParameter.
   * @param {Object} args Object containing the parameter properties
   */
  constructor(args) {
    this._id = args._id;
    this._model = args.model;
    this.component = args.component;
    this.description = args.description;
    this.internalSearch = args.internalSearch;
    this.label = args.label;
    this.placeholder = args.label;
    this.taggable = args.taggable;
    this.testId = `fr-field-${args._id}`;
    this.type = args.type;

    if (args.request) {
      this.request = args.request;
      this.request.fetchRequestData = (term) => this.request.fetch(this.request, term);
      this.request.mutateRequestData = (data) => this.request.mutation(data, this.request.attribute);
    } else if (args.enums) {
      this.enums = ReportParameter.enumMutation(args.enums);
    } else if (args.type === 'multiselect' || args.type === 'select') {
      this.options = [];
    }
  }

  get model() {
    return this._model;
  }

  set model(value) {
    this._model = value;
  }

  /**
   * @description Method to handle the data for a select / multiselect field.
   * @param {Array} data Data handled by the dropdown options
   * @returns {Array} Filtered array if field type is select or multiselect
   */
  selectOptionsDataHandler(data) {
    if (data && this.model) {
      if (this.type === 'select') {
        // The select component requires a selected value to be part of the
        // overall list of options and if it is not present, the selected value
        // will not be displayed.  This can happen when we fetch data when
        // searching within a field and selecting a value that is not in the
        // original list. After selection, the field will query for the original
        // list, which will not contain the selected value, and will make the
        // selected value disappear. To prevent this, we add the selected value
        // to the beginning of the list if it is not included in the response.
        if (!data.includes(this.model)) return [this.model, ...data];
      }
      if (this.type === 'multiselect') {
        // Filters out any selected values from the search results
        return data.filter((item) => !this.model.includes(item));
      }
    }

    return data;
  }

  /**
   * @description Method to format the payload for the API request.
   * @returns {Object} containing the parameter payload
   */
  payload() {
    let payload = this.model;
    if (this.type === 'date') payload = ReportParameter.dateToISO(payload);
    return { [this._id]: payload };
  }

  /**
   * @description Setter for the select component options which
   * can be triggered on the "@tag" event on the <FrField> component.
   * It checks if the value is already present in the enums,
   * request, or options arrays and adds it if it is not present.
   * @param {String} val The value to be added to the dropdown options
   */
  selectOptionsSetter(val) {
    const { enums, request, options } = this;
    const enumsText = enums ? enums.map(({ text }) => text.toLowerCase()) : [];

    if (enums && !enumsText.includes(val.toLowerCase())) {
      enums.unshift(ReportParameter.enumFormat(val));
    } else if (request && !request.data.includes(val)) {
      request.data.unshift(val);
    } else if (options && !options.includes(val)) {
      this.options.unshift(val);
    }

    this.model = val;
  }

  /**
   * @description Method to convert a date string to ISO format.
   * @param {String} date date string from the datepicker component
   * @returns {String} ISO formatted date string
   */
  static dateToISO(date) {
    return Date.parse(date) ? new Date(date).toISOString() : date;
  }

  /**
   * @description Method to format the enums that are received from the API.
   * @param {String} value use for the label and value of the enum
   * @returns {Object} containing the text and value of the enum
   */
  static enumFormat(value) {
    return { text: value, value };
  }

  /**
   * @description Formats the enums array that is passed to the FrField component.
   * @param {Array} enums Array of enums to be formatted
   * @returns {Array} containing the text and value of the enums
   */
  static enumMutation(enums) {
    return enums.map(({ name: text, value }) => ({ text, value }));
  }
}
