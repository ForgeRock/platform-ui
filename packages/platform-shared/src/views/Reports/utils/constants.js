/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

/**
 * Report viewer group name
 */
export const REPORT_VIEWER_GROUP = 'report_viewer';

/**
 * Used for mapping field type names that come from the API,
 * to the field type names that the <FrField> component consumes.
 */
export const FIELD_TYPE_MAP = {
  array: 'multiselect',
  boolean: 'boolean',
  date: 'date',
  float: 'number',
  integer: 'number',
  string: 'string',
};
