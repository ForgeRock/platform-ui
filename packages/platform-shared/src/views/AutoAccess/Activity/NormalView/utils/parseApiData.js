/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { colorsHex, colorsReadable } from './colors';

export default function parseApiData(response) {
  const categories = {};
  const categoryHasOther = {};
  const categoryNames = [...new Set(response.result.map((item) => item.category))];
  categoryNames.forEach((category) => {
    let categoryValue = {};
    let otherCount = 0;
    let otherCategoryData = {};
    categories[category] = [];
    categoryHasOther[category] = false;

    /**
     * This reorganizing of the API data is needed since the data provided is formatted in a way needed by the UI.
     *
     * The data is flat and need to be organized into a usable object structure.
     * It needs to be manipulated by merging values of 'other', and null together into a single field.
     * The field of 'Others' also need to be converted to have the label 'other' and be sorted last
     */
    response.result.forEach((item) => {
      if (item.category === category) {
        // 'other' or null
        if (item.value === 'other' || !item.value) {
          otherCount += parseInt(item.count, 10);
          if (!categoryHasOther[category]) {
            otherCategoryData = {
              label: 'unknown',
              value: otherCount,
              valueFormatted: otherCount,
            };
            categoryHasOther[category] = true;
          } else {
            otherCategoryData.value = otherCount;
            otherCategoryData.valueFormatted = otherCount;
          }
        } else {
          // Handle all normal categories, except 'Others' which gets renames to 'other'
          categoryValue = {
            label: item.value === 'Others' ? 'other' : item.value,
            value: parseInt(item.count, 10),
            valueFormatted: item.count,
          };
          categories[item.category].push(categoryValue);
        }
      }
    });

    // We have found 'other' or null and need to add the manipulated field to the main array
    if (categoryHasOther[category]) {
      categories[category].push(otherCategoryData);
    }

    // Sort categories based on numerical 'value', but leave 'other' sorted last
    Object.keys(categories).forEach((cat) => {
      categories[cat].sort((a, b) => {
        const aVal = parseInt(a.value, 10);
        const bVal = parseInt(b.value, 10);
        if (a.label === 'other') { return 1; }
        if (b.label === 'other') { return -1; }
        if (aVal < bVal) { return 1; }
        if (aVal > bVal) { return -1; }
        return 0;
      });

      // Set the color values - hex for the Donut Chart, readable for the data table
      let colorIndex = 0;
      categories[cat].forEach((item) => {
        item.color = colorsHex[colorIndex] || colorsHex[4];
        item.colorReadable = colorsReadable[colorIndex] || colorsReadable[4];
        colorIndex += 1;
      });
    });
  });

  return categories;
}
