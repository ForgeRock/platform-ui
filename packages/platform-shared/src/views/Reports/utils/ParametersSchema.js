/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { FIELD_TYPE_MAP } from './constants';
import DATA_CONTROLLER from './DataController';
import ReportParameter from './ReportParameter';
import store from '@/store';

/**
  * @description Defines the schema for the report parameters.
  * @param {Object} reportConfig Report template config parameters
  * @param {Boolean} isPrePackagedReport Determines if the report is pre-packaged (out-of-the-box)
  */

export default function ParametersSchema(reportConfig, isPrePackagedReport) {
  const isPlatformAdmin = store.state.SharedStore.currentPackage === 'admin';
  const { parameters: configParameters } = reportConfig;
  const configParameterKeys = configParameters ? Object.keys(configParameters) : [];

  if (configParameterKeys.length) {
    return configParameterKeys.map((_key) => {
      const {
        attribute,
        description = '',
        entity = '',
        enum: enums,
        label = _key,
        source,
        type = 'string',
      } = configParameters[_key];
      const SERVICE = Object.keys(DATA_CONTROLLER).find((key) => (
        Object.keys(DATA_CONTROLLER[key]).includes(entity)
      ));
      const entityOverrides = SERVICE ? DATA_CONTROLLER[SERVICE][entity] : {};
      const attributeOverrides = entityOverrides[attribute] || {};
      const { canQuery, endUserCanFetch } = entityOverrides;
      const hasEnums = !!enums?.length;
      const hasTaggableOverride = attributeOverrides.taggable === true || entityOverrides.taggable === true;
      const isSelectField = type === 'string' && (hasEnums || source === 'datasource');
      const canFetch = !!(SERVICE && !hasEnums && (isPlatformAdmin || endUserCanFetch));
      const fieldType = isSelectField ? 'select' : FIELD_TYPE_MAP[type];
      const fieldIsADropdown = fieldType === 'select' || fieldType === 'multiselect';
      const internalSearch = fieldIsADropdown && (canFetch === false || canQuery === false);
      const isUnmappedEntity = !SERVICE && source === 'datasource' && !hasEnums;
      const sourceIsBasicOrUserProvided = source === 'basic' || source === 'user_provided';
      const taggable = fieldIsADropdown && (sourceIsBasicOrUserProvided || !isPlatformAdmin || hasTaggableOverride || isUnmappedEntity);

      let model = '';
      let component = 'FrField';

      if (fieldType === 'boolean') model = false;
      if (fieldType === 'multiselect') model = [];
      if (fieldType === 'number') model = null;
      // The only real hard-coded exception here is for original out-of-the-box
      // reports needing to map to a custom component for the date range.
      if (isPrePackagedReport && _key === 'startDate') component = 'FrTimeframeField';
      if (isPrePackagedReport && _key === 'endDate') component = null;

      return {
        [_key]: new ReportParameter({
          _id: _key,
          component,
          description,
          internalSearch,
          label,
          model,
          taggable,
          type: fieldType,
          // conditional properties
          ...(hasEnums && { enums }),
          ...(canFetch && {
            request: {
              attribute,
              entity,
              data: [],
              fetch: DATA_CONTROLLER[SERVICE].fetch,
              mutation: (data, attr) => data.map((obj) => obj[attr] || obj),
              ...entityOverrides,
              ...attributeOverrides,
            },
          }),
        }),
      };
    }).reduce((a, c) => ({ ...a, ...c }), {}); // Merges the array of objects into a single object
  }
  return {};
}
