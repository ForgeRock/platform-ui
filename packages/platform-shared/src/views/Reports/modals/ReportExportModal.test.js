/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { mount } from '@vue/test-utils';
import { findByTestId } from '@forgerock/platform-shared/src/utils/testHelpers';
import i18n from '@/i18n';
import ReportExportModal from './ReportExportModal';

describe('Report Export modal component', () => {
  function setup(props) {
    return mount(ReportExportModal, {
      i18n,
      propsData: {
        isTesting: true,
        ...props,
      },
    });
  }

  let wrapper;

  describe('@renders', () => {
    beforeEach(() => {
      wrapper = setup();
    });

    it('ensures that the status of the modal is in loading state on mount', () => {
      const spinner = findByTestId(wrapper, 'spinner-large');
      expect(spinner.exists()).toBe(true);
    });

    it('ensures that the status of the modal changes to the download state if the status prop equals to "download"', () => {
      wrapper = setup({ status: 'download' });

      const spinner = findByTestId(wrapper, 'spinner-large');
      expect(spinner.exists()).toBe(false);
    });

    it('ensures that the status of the modal changes to the download state if the status prop equals to "downloading"', () => {
      wrapper = setup({ status: 'downloading' });

      const spinner = findByTestId(wrapper, 'spinner-large');
      expect(spinner.exists()).toBe(false);
    });

    it('ensures that an error image is displayed if the status prop equals to "error"', () => {
      const spinner = findByTestId(wrapper, 'spinner-large');
      expect(spinner.exists()).toBe(true);

      wrapper = setup({ status: 'error' });

      const errorImage = findByTestId(wrapper, 'fr-export-error-image');
      const spinnerAfterError = findByTestId(wrapper, 'spinner-large');
      expect(errorImage.exists()).toBe(true);
      expect(spinnerAfterError.exists()).toBe(false);
    });

    it('ensures that the download button is displayed if the status prop equals to "download"', () => {
      const downloadButton = findByTestId(wrapper, 'fr-history-export-download-button');
      expect(downloadButton.exists()).toBe(false);

      wrapper = setup({ status: 'download' });
      const downloadButtonAfterStatusChange = findByTestId(wrapper, 'fr-history-export-download-button');
      expect(downloadButtonAfterStatusChange.exists()).toBe(true);
    });

    it('ensures that the download button is displayed if the status prop equals to "downloading"', () => {
      const downloadButton = findByTestId(wrapper, 'fr-history-export-download-button');
      expect(downloadButton.exists()).toBe(false);

      wrapper = setup({ status: 'downloading' });
      const downloadButtonAfterStatusChange = findByTestId(wrapper, 'fr-history-export-download-button');
      expect(downloadButtonAfterStatusChange.exists()).toBe(true);
    });

    it('disables the download button when the status prop is equal to "downloading"', () => {
      wrapper = setup({ status: 'downloading' });
      const downloadButton = findByTestId(wrapper, 'fr-history-export-download-button');
      expect(downloadButton.attributes().disabled).toBe('disabled');
    });

    it('emits "download-report" when the download button is clicked', async () => {
      wrapper = setup({
        data: {
          runId: 'job_0123',
        },
        fileType: 'JSON',
        status: 'download',
      });
      const downloadButton = findByTestId(wrapper, 'fr-history-export-download-button');

      await downloadButton.trigger('click');
      expect(wrapper.emitted()['download-report'][0]).toEqual(['JSON', { runId: 'job_0123' }]);
    });
  });
});
