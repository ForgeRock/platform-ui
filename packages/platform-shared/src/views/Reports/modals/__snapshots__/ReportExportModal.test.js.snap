// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Export modal component @renders ensures that the status of the modal changes to the download state if the status prop equals to "download": console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (5)",
    "
",
    " at <BImg",
    "key=2",
    "class="m-4"",
    "height="160"",
    " ...",
    ">",
    "
",
    " at <BCol",
    "md="10"",
    "offset-md="1"",
    "class="text-center"",
    ">",
    "
",
    " at <BRow>",
    "
",
    " at <BContainer>",
    "
",
    " at <Transition",
    "onBeforeEnter=fn<bound onBeforeEnter>",
    "onEnter=fn<bound onEnter>",
    "onAfterEnter=fn<bound onAfterEnter>",
    " ...",
    ">",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "status="download"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (4)",
    "
",
    " at <ButtonWithSpinner",
    "key=3",
    "class="mb-4"",
    "data-testid="fr-history-export-download-button"",
    " ...",
    ">",
    "
",
    " at <BCol",
    "md="10"",
    "offset-md="1"",
    "class="text-center"",
    ">",
    "
",
    " at <BRow>",
    "
",
    " at <BContainer>",
    "
",
    " at <Transition",
    "onBeforeEnter=fn<bound onBeforeEnter>",
    "onEnter=fn<bound onEnter>",
    "onAfterEnter=fn<bound onAfterEnter>",
    " ...",
    ">",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "status="download"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) (3)",
    "
",
    " at <ButtonWithSpinner",
    "key=3",
    "class="mb-4"",
    "data-testid="fr-history-export-download-button"",
    " ...",
    ">",
    "
",
    " at <BCol",
    "md="10"",
    "offset-md="1"",
    "class="text-center"",
    ">",
    "
",
    " at <BRow>",
    "
",
    " at <BContainer>",
    "
",
    " at <Transition",
    "onBeforeEnter=fn<bound onBeforeEnter>",
    "onEnter=fn<bound onEnter>",
    "onAfterEnter=fn<bound onAfterEnter>",
    " ...",
    ">",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "status="download"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`Report Export modal component @renders ensures that the status of the modal is in loading state on mount: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (2)",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (2)",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation OPTIONS_BEFORE_DESTROY) \`beforeDestroy\` has been renamed to \`beforeUnmount\`.",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_SCOPED_SLOTS) vm.$scopedSlots has been removed. Use vm.$slots instead.
  Details: https://v3-migration.vuejs.org/breaking-changes/slots-unification.html",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (3)",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) Functional component <BButtonClose> should be defined as a plain function in Vue 3. The "functional" option has been removed. NOTE: Before migrating to use plain functions for functional components, first make sure that all async components usage have been migrated and its compat behavior has been disabled.
  Details: https://v3-migration.vuejs.org/breaking-changes/functional-components.html",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) (2)",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) vm.$listeners has been removed. In Vue 3, parent v-on listeners are included in vm.$attrs and it is no longer necessary to separately use v-on="$listeners" if you are already using v-bind="$attrs". (Note: the Vue 3 behavior only applies if this compat config is disabled)
  Details: https://v3-migration.vuejs.org/breaking-changes/listeners-removed.html",
    "
",
    " at <BButtonClose",
    "onClick=fn<bound onClose>",
    "ref="close-button"",
    "content="&times;"",
    " ...",
    ">",
    "
",
    " at <Transition",
    "onBeforeEnter=fn<bound onBeforeEnter>",
    "onEnter=fn<bound onEnter>",
    "onAfterEnter=fn<bound onAfterEnter>",
    " ...",
    ">",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (2)",
    "
",
    " at <BContainer>",
    "
",
    " at <Transition",
    "onBeforeEnter=fn<bound onBeforeEnter>",
    "onEnter=fn<bound onEnter>",
    "onAfterEnter=fn<bound onAfterEnter>",
    " ...",
    ">",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (3)",
    "
",
    " at <BButton",
    "textContent="Close"",
    "onClick=fn<bound onOk>",
    "ref="ok-button"",
    " ...",
    ">",
    "
",
    " at <Transition",
    "onBeforeEnter=fn<bound onBeforeEnter>",
    "onEnter=fn<bound onEnter>",
    "onAfterEnter=fn<bound onAfterEnter>",
    " ...",
    ">",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "appear" with v-bind value \`false\` will render appear="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <Transition",
    "onBeforeEnter=fn<bound onBeforeEnter>",
    "onEnter=fn<bound onEnter>",
    "onAfterEnter=fn<bound onAfterEnter>",
    " ...",
    ">",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "persisted" with v-bind value \`false\` will render persisted="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <Transition",
    "onBeforeEnter=fn<bound onBeforeEnter>",
    "onEnter=fn<bound onEnter>",
    "onAfterEnter=fn<bound onAfterEnter>",
    " ...",
    ">",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (4)",
    "
",
    " at <BVTransition",
    "noFade=false",
    ">",
    "
",
    " at <BModal",
    "data-testid="export-modal"",
    "id="export-modal"",
    "ok-only=""",
    " ...",
    ">",
    "
",
    " at <ReportExportModal",
    "isTesting=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_EVENT_EMITTER) vm.$on/$once/$off() have been removed. Use an external event emitter library instead.
  Details: https://v3-migration.vuejs.org/breaking-changes/events-api.html",
  ],
]
`;
